### PR TITLE
composer, phpunit, few fixes

### PR DESCRIPTION
**Pull request by mjanda**
Ok, this will probably need some review.

I've added composer.json and it seems to be working (with repository override)

```
require: {
    "adriangibbons/php-FIT-File-Analysis": "dev-master"
},
...
"repositories": [{
    "type": "vcs",
    "url": "https://github.com/mjanda/php-FIT-File-Analysis"
}]
```

Other than that, I've:
- added running dynamics fields (vertical_oscillation and stance_time)
- changed read_header to unpack crc only when header length > 12
- added check `} else if (isset($this->types[$field_defn['base_type']])) {` - it was throwing `undefined offset 12` with some activities and `error_reporting(E_ALL)` - I'll try to find out what type is field 12, but it should probably ignore unknown fields anyway
- added basic phpunit (just tries to load any files in test_files directory and check if timestamp > 0)
- changed demo files to use current class without copying it elsewhere
- changed demo files to use paths relative to them, so they can be invoked from different directory (I was calling them from command line just to see if they error or not)
- and ehm ehm...forgot that I have php.fmt enabled and it reformatted the source to something closer to PSR standard - if this is a problem, I can go back and re-do those changes on original source

Due to namespace changes, it's not drop in upgrade, so some documentation change will be needed too. But I think some native speaker would be better suited for that.

### CODE REVIEW

**Comment by mjanda** [comment]
Closing until I fix one problem with fix_data.

**Comment by mjanda** [comment]
Ok, I've fixed the problem.
interpolate_missing_data() would get stuck inside infinite loop if missing value would be outside $array range.

Hey, Fenix 2 really can produce shitty files :poop:

I've changed the test case to test with fix_data=all

I guess it would be useful to collect various buggy files in test_files.

**Comment by adriangibbons** [comment]
Thanks @mjanda! I will look at this in the next week or so as I am travelling around at the moment. I will get to it and your contribution appreciated!

**Comment by mjanda** [comment]
Ok, I'm writing some basic FIT parser in Go now, mostly using your class as reference and I've realized some of my changes in this PR are clearly wrong, so consider it buggy. On the other hand, I think original behavior is wrong too (for example timestamp is used as index before it's encountered, so it's either 0 or previous value).

If you want to cherry pick composer related changes, feel free. Otherwise, wait until I'll get back to civilization and I'll try to take another look and fix it properly.

**Comment by adriangibbons** [comment]
Thanks @mjanda I've captured the timestamp issue as #6 - if there are any more bugs in my code you have found then let me know. I am actually on the road myself and don't really have the time/resources to look at this properly for another week or so either. Cheers

File: php-FIT-File-Analysis.php [hunk 1]
```diff
@@ -583,32 +585,33 @@ function __construct($file_path, $options=NULL) {
...
-			'V1data_size/'.
-			'C4data_type/'.
-			'v1crc', substr($this->file_contents, $this->file_pointer, $header_size - 1)
-		);
+        $header_fields = 'C1protocol_version/' .
+                         'v1profile_version/' .
+                         'V1data_size/' .
+                         'C4data_type';
+        if ($header_size > 12) {
+            $header_fields .= '/v1crc';
```
Comment (mjanda) [inline php-FIT-File-Analysis.php:613]: You allow 12 byte headers, but try to extract 14 bytes from them

File: php-FIT-File-Analysis.php [hunk 2]
```diff
@@ -616,11 +619,19 @@ private function read_data_records() {
...
 		$local_mesg_type;
-		
+
+        // value signifies invalid data point for that field size
+        $invalid_values = [
+            1=>255,
+            2=>65535,
+            4=>2147483647
+        ];
+
```
Comment (mjanda) [inline php-FIT-File-Analysis.php]: At least some of the fields use those values to mean "invalid value" - missing latitude/longitude, vertical oscillation/stance time when standing etc...so we won't even save those values and interpollate them when fixing

File: php-FIT-File-Analysis.php [hunk 3]
```diff
@@ -668,25 +679,28 @@ private function read_data_records() {
...
+                                    $field_value = (unpack($this->types[$field_defn['base_type']], substr($this->file_contents, $this->file_pointer, $field_defn['size']))['tmp'] / $this->data_mesg_info[$this->defn_mesgs[$local_mesg_type]['global_mesg_num']]['field_defns'][$field_defn['field_definition_number']]['scale']) - $this->data_mesg_info[$this->defn_mesgs[$local_mesg_type]['global_mesg_num']]['field_defns'][$field_defn['field_definition_number']]['offset'];
+                                    if (isset($invalid_values[$field_defn['size']]) && $field_value != $invalid_values[$field_defn['size']]) {
+									   $this->data_mesgs[$this->data_mesg_info[$this->defn_mesgs[$local_mesg_type]['global_mesg_num']]['mesg_name']][$this->data_mesg_info[$this->defn_mesgs[$local_mesg_type]['global_mesg_num']]['field_defns'][$field_defn['field_definition_number']]['field_name']][$this->timestamp] = $field_value;
+                                    }
 								}
-								
-								else {
+
+								else if (isset($this->types[$field_defn['base_type']])) {
```
Comment (mjanda) [inline php-FIT-File-Analysis.php]: gave me invalid offset otherwise

File: php-FIT-File-Analysis.php [hunk 4]
```diff
@@ -792,17 +806,33 @@ private function fix_data($options) {
...
 		}
 	}
-	
+
 	/*
 	 * For the missing keys in the data, interpolate using values either side and insert as necessary.
 	 */
 	private function interpolate_missing_data(&$missing_keys, &$array){
+        if (!is_array($array)) {
+            // can't interpollate from non-array
```
Comment (mjanda) [inline php-FIT-File-Analysis.php]: Got null or false in &$array with some files

File: php-FIT-File-Analysis.php [hunk 5]
```diff
@@ -792,17 +806,33 @@ private function fix_data($options) {
...
 		for($i=0; $i<count($missing_keys); ++$i) {
 			if($missing_keys[$i] !== 0) {
+                // interpollating outside recorded range is impossible - use edge values instead
+                if ($missing_keys[$i] > $max_key) {
+                    $array[$missing_keys[$i]] = $array[$max_key];
+                    continue;
+                } else if ($missing_keys[$i] < $max_key) {
+                    $array[$missing_keys[$i]] = $array[$min_key];
+                    continue;
+                }
```
Comment (mjanda) [inline php-FIT-File-Analysis.php:859]: When there are missing values before/after first/last valid value, it would ho into infinite loop

**Comment by mjanda** [comment]
Redid the PR on originally formatted file. Kept back some low-priority changes (running dynamics etc.) to make the review more manageable.

**Comment by adriangibbons** [comment]
Hi @mjanda, thanks for your patience. Life has been very hectic lately and I haven't had much chance to look at my project for some time!

I'm working my way through your individual commits and adding them to my project manually rather than accepting the PR as I had made some other changes and conflicts etc.

So far, I have incorporated:
- [changed read_header to unpack crc only when header length > 12](https://github.com/adriangibbons/php-FIT-File-Analysis/commit/90f78a5964cce29f35315fbc29f24b959571cac1)
- [enhancements to interpolate_missing_data()](https://github.com/adriangibbons/php-FIT-File-Analysis/commit/fef948896c156b451b58b988696c7225ada74550)

It would be very useful to have a copy of the Fenix 2 (or any other devices) files to test and investigate the issues you found. You could share via GitHub or email me (address in the comments at top of PHP file) if you prefer

I have removed the hard dependency on the PHP Trader PECL extension by adding an alternate Simple Moving Average algorithm in the class itself (though it is slower). Does this have implications for the composer.json file?

``` JSON
"require": {
    "ext-trader": "*"
},
```

I will incorporate more of your good work as I am able to find the time to work through it.

Hope your project creating a FIT parser in Go is making good progress!

Regards, Adrian.

**Comment by mjanda** [comment]
If you've added your own SMA function, you can drop that ext-trader from composer.json

**Comment by adriangibbons** [comment]
Thanks for the FIT files. I've added my own SMA function, in case the PHP Trader extension is not loaded on the system - the preference would be to still use PHP Trader if it's available as it is much faster, presumably as it is compiled.

**Comment by adriangibbons** [comment]
Regarding:

> added check `} else if (isset($this->types[$field_defn['base_type']])) {` - it was throwing `undefined offset 12` with some activities and `error_reporting(E_ALL)` - I'll try to find out what type is field 12, but it should probably ignore unknown fields anyway

There is no base type of 12 (decimal). As per the document in the standard, _D00001275 Flexible & Interoperable Data Transfer (FIT) Protocol Rev 1.7.pdf_, Table 4-6. FIT Base Types and Invalid Values, I have an array that stores the base types:

``` php
private $types = array(
    0   => 'Ctmp',  // enum
    1   => 'ctmp',  // sint8
    2   => 'Ctmp',  // uint8
    131 => 'Stmp',  // sint16
    132 => 'vtmp',  // uint16
    133 => 'ltmp',  // sint32
    134 => 'Vtmp',  // uint32
    7   => 'Ctmp',  // string
    136 => 'ftmp',  // float32
    137 => 'dtmp',  // float64
    10  => 'Ctmp',  // uint8z
    139 => 'vtmp',  // uint16z
    140 => 'Vtmp',  // uint32z
    13  => 'Ctmp',  // byte
);
```

Hence the exception when `base_type` is 12. So I guess technically it is not a valid FIT message/file and I am not sure whether I should just add an `isset()` as you propose and cater for this scenario or not. Hmmmm.

**Comment by mjanda** [comment]
Well, current state is that you either feed this possibly invalid data to unpack() or throw exception - depending on user's error reporting configuration - basically undefined behavior. Imho, knowingly skipping it is a better solution.

**Comment by adriangibbons** [comment]
Agreed, and I have incorporated a [check for a valid base type](https://github.com/adriangibbons/php-FIT-File-Analysis/commit/9602bfd44b54940da26a3634da1f770832ae6e53#diff-736e2bc29e1e0e311649c2e7ad1526c8) just now slightly higher up in the code than you had suggested. I have an idea to fix the **timestamp** race condition and plan to implement that today.

**Comment by adriangibbons** [comment]
[timestamp race condition fix](https://github.com/adriangibbons/php-FIT-File-Analysis/commit/2bad69e5c6de10398b970cee1fe5dfc38d911801) has now been merged into the master branch. Now record fields are collected in a temporary array before being added to the public data_mesgs array so it does not matter which order they are received, as long as they have timestamp - thanks for identifying this issue!

**Comment by adriangibbons** [comment]
Hi @mjanda, I've worked my way through many of the changes you proposed and have included them manually as I am still getting to grips with GitHub.

I think there are just three things remaining to incorporate, which are captured as issues:
1. Composer package & submit to packagist #4 
2. Make code compliant with the PHP-FIG Coding Style Guide (PSR-2) #7 
3. Add PHPUnit tests #9 

Therefore I am going to close this PR for now as my master branch is now 28 commits ahead of it and I have incorporated most of your changes.

Thanks again, Adrian.

**Pull request closed without merging**